### PR TITLE
feat: [DHIS2-14775][DHIS2-14729] Render data entry plugins

### PR DESCRIPTION
**Pull request by eirikhaugstulen**

### CODE REVIEW

**Comment by eirikhaugstulen** [comment]
Lockfile seems to not respond to the resolutions we set because we're running an alpha version of the app-platform

File: src/core_modules/capture-core/components/DataEntries/common/TEIAndEnrollment/useMetadataForRegistrationForm/hooks/useEnrollmentFormFoundation.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    cachedTrackedEntityAttributes,
+    dataEntryFormConfig,
+    configIsFetched,
+    program,
+    locale,
+}: Props) => {
+    const { data: enrollment, isLoading, error } = useIndexedDBQuery(
+        ['enrollmentForm', program?.id],
+        () => buildEnrollmentForm({
+            // $FlowFixMe - Flow does not understand that the values are not null here
```
Comment (simonadomnisoru) [inline src/core_modules/capture-core/components/DataEntries/common/TEIAndEnrollment/useMetadataForRegistrationForm/hooks/useEnrollmentFormFoundation.js:35]: Hey, there are quite a few ` // $FlowFixMe` in this file. Can anything be done to improve this?

File: src/core_modules/capture-core/metaDataMemoryStoreBuilders/programs/factory/enrollment/EnrollmentFactory.js [hunk 2]
```diff
@@ -129,8 +134,24 @@ export class EnrollmentFactory {
     ) {
         // $FlowFixMe
         await cachedProgramTrackedEntityAttributes.asyncForEach(async (trackedEntityAttribute) => {
-            const element = await this.dataElementFactory.build(trackedEntityAttribute);
-            element && section.addElement(element);
+            if (trackedEntityAttribute?.id === 'plugin') {
```
Comment (simonadomnisoru) [inline src/core_modules/capture-core/metaDataMemoryStoreBuilders/programs/factory/enrollment/EnrollmentFactory.js:137]: The string `'plugin'` is used a couple of times across the PR. Does it make sense to move to a const in DataEntryPlugin.const.js ?

File: src/core_modules/capture-core/metaDataMemoryStoreBuilders/trackedEntityTypes/factory/TrackedEntityType/TeiRegistrationFactory.js [hunk 3]
```diff
@@ -79,12 +84,66 @@ export class TeiRegistrationFactory {
...
+                        await trackedEntityAttribute.fieldMap.asyncForEach(async (field) => {
+                            const dataElement = await this.dataElementFactory.build(field);
+                            dataElement && element.addField(field.IdFromPlugin, dataElement);
+                        });
+
+                        element && section.addElement(element);
+                    } else {
+                        const element = await this.dataElementFactory.build(trackedEntityAttribute);
+                        element && section.addElement(element);
+                    }
```
Comment (simonadomnisoru) [inline src/core_modules/capture-core/metaDataMemoryStoreBuilders/trackedEntityTypes/factory/TrackedEntityType/TeiRegistrationFactory.js:137]: This is the same logic as in `EnrollmentFactory.js`. Can it be extracted to a commonplace and reused?

**Review by simonadomnisoru — COMMENTED** [review]
Great job on this feature @eirikhaugstulen! I added a few small comments that I hope you can have a look at. Thanks!

**Review by JoakimSM — COMMENTED** [review]
Looks really good @eirikhaugstulen 👍 Let's have a chat tomorrow.

File: src/core_modules/capture-core/components/D2Form/DataEntryPlugin/DataEntryPlugin.container.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+    // Expanding iframe height temporarily to fit content - LIBS-487
+    useEffect(() => {
+        const iframe = document.querySelector('iframe');
+        if (iframe) iframe.style.height = '500px';
+    }, []);
+
+
+    // Formatted metadata is needed to remove the underscore from the keys
+    const formattedMetadata = useMemo(() => Array.from(fieldsMetadata.entries())
```
Comment (JoakimSM) [inline src/core_modules/capture-core/components/D2Form/DataEntryPlugin/DataEntryPlugin.container.js:32]: We also have to convert option sets. Adding an option set right now causes a circular structure error.

Additionally, I think we should not pass on the app-ids, so we don't encourage using those (won't work cross instance)

File: src/core_modules/capture-core/components/D2Form/DataEntryPlugin/DataEntryPlugin.types.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+import type { DataElement } from '../../../metaData';
+
+
+type FieldValueOptions = {|
+    valid?: boolean,
+    touched?: boolean,
+    error?: string,
+|}
+
+export type MetadataByPluginId = { [id: string]: DataElement }
```
Comment (JoakimSM) [inline src/core_modules/capture-core/components/D2Form/DataEntryPlugin/DataEntryPlugin.types.js:12]: Let's define a new type for the converted data element object we created.

File: src/core_modules/capture-core/components/D2Form/FormBuilder/FormBuilder.component.js [hunk 6]
```diff
@@ -500,6 +531,31 @@ export class FormBuilder extends React.Component<Props> {
         }, []);
     }
 
+    renderPlugin = (
+        field: Object,
```
Comment (simonadomnisoru) [inline src/core_modules/capture-core/components/D2Form/FormBuilder/FormBuilder.component.js:535]: I believe the type can be improved to `field: FieldConfig` if `field.props` is checked for null or undefined [below](https://github.com/dhis2/capture-app/pull/3222/files#diff-34437e7d08816ce5e6cc3c54623ac10f394f1c42f131e84664f15a12902b3e7dR545). WDYT? Thanks!

**Review by superskip — COMMENTED** [review]
Just added a small comment on a detail, did not make a proper review.

**Comment by eirikhaugstulen** [comment]
We're reverting back to displaying an error message for now until the linked issues are merged into app-platform. Thanks for the reviews @simonadomnisoru & @JoakimSM. Should be fixed now. Will change back to plugin component in https://dhis2.atlassian.net/browse/DHIS2-15475.

**Review by JoakimSM — APPROVED** [review]

**Review by geethaalwan — APPROVED** [review]
Tested successfully on 2.41 version

**Pull request merged**